### ISSUE TITLE
[TW-238] new attribute over:<date> as end-of-recurrence-period

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#281]
_David Patrick on 2012-08-30T19:18:46Z says:_ 

 UPDATED!

TL:DR
Recurrent parent tasks are over when it's over:<date> 
until:<date>, sched:<date> and wait:<date> should apply only to child tasks, and should be defined relative to the child due:<date>.

======================================================

I know what you're thinking; "what the h. e. double-hockey-sticks?? The until: attribute works _fine_ with recur:.. it was DESIGNED for it!"

Well that's true. until:<date> was originally just a way to indicate a stop-date for recurring tasks. After a feature request, and some more "typing" by Paul, the until:<date> was extended as an auto-delete end-date for _any_ task, and that works well.. a little TOO well.

Here's the problem;

The four date parameters can now form a full time-envelope around a task; from when visible, actionable, deliverable and no-longer-do-able, deleted. My personal goal, one that is likely shared by anyone who typically carries more than 300 TODOs, is to stop seeing just a forest, and actually look at some trees. 

I believe it's possible to set up a fine-grained, highly-detailed task list that covered everything, personal routine, work tasks, life-goals, books-to-read, diet, grooming, dogwalking, _you name it_, and add 1000+ undated TODOs (but _with_ context and duration) and to have taskwarrior generate a list that is actionable and _not_ overwhelming. 

Taskwarrior can do this by a) limiting the overall time-horizon and hiding non-actionable tasks, b) breaking the listing into logical sections, sorting on logical keys. c) coherent color themes d) tweaking the time-envelope of every task, especially the routine ones, so that it appears no earlier than you need it (wait:) and it is deleted as soon as it is no longer actionable (until). 
If your list for today was jam-packed with detail, highlights listed for tomorrow, and only key items shown beyond that, it wouldn't matter how many tasks were in the system, you could limit your perspective to the immediate horizon of highest-urgency actionable tasks.

The problem arises with "d)", in trying to minimally reveal tasks, like a pop-up-book, dubbed elsewhere as a "rolling window of opportunity". To set an until:date for a recurring task applies to the whole recurrence period (as it always has) and can not be applied to each recurring task as a "kill-date". Because of this, low (or no) priority daily activities (or contexts) can be listed and recur, but they can not clean up after themselves, leaving the time-envelope open on one side, and trivial tasks left littered in the list, instead of auto-deleting by 11am. The _last_ thing you need is to add a "clean up routine activity cruft from task manager" task to your list, recurring or otherwise.

I've scratched my head and looked at this thing sideways for a while, and I don't see a workaround, other than a new date parameter that specifically stops recurrences, leaving the existing "until:" date to be used as a self-destruct-date for each new recurrence (and still for regular tasks too). Without this, untended, uncompleted, recurring, no-longer-actionable tasks will just pile up.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2012-08-30T19:18:46Z
Modified: 2015-09-01T23:47:53Z
```

**Comment by taskwarrior** [comment]
_David Patrick on 2013-01-05T18:25:34Z says:_ 

 While this is an important function for fully automating the life-cycle of recurrent tasks, as described above, the suggested name (recur-until) won't fly, because it's ambiguous till five characters with the existing "recur:<period>".. but it _could_ be called *end-recur:<date>*.. or something..

**Comment by taskwarrior** [comment]
_David Patrick on 2013-05-04T18:45:16Z says:_ 

 Further tweaks to the current, due-for-overhaul, recurrence model, especially something as complicated as end-recurrences, might be sort of futile, so I had an idea for an alternate approach to close the "time-envelope"; (magic noise) script, hooks and special tag +expire !

This new special tag could be added to trigger a clean-up script, automatically marking such tasks as deleted. When will it be deleted? That should depend on the task is due, should be configurable, and should have sane defaults. After thinking about it, I think the expiration should take place on due-date+1 (the day after the due date) or recurrence-period+1 (that is one-week-after, for recur:weekly, one-month-after for recur:monthly, etc) allowing some retrospect, cleaning up, but not yanking it out from under you.  I'm trying to imagine how (if at all) it might affect a task with no due:date. 

This should be optional behavior, but it would finally allow the use of fine-grained recurrences without the cruft.

**Comment by taskwarrior** [comment]
_Benjamin Weber on 2014-05-26T20:52:49Z says:_ 

 ```
task add Clean the windows recur:<period>:[<start>]:[<end>]
```
(<start> and <end> are optional. <end> requires <start> to be supplied.)

The parent could hold the <start> and <end> interval information.
Is that similar to <limit>?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-07-02T02:28:59Z says:_ 

 I cannot think of a use case where you want to shut off recurence and retain the template task.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-07-02T02:29:12Z says:_ 

 No use case.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-03-28T15:23:34Z says:_ 

 Okay, then what is the use case?

**Comment by taskwarrior** [comment]
_David Patrick on 2015-03-29T05:23:13Z says:_ 

 Ok, the use-case is to overcome a key weakness in the recurrence model, that wait, sched and until values are not translated to the child task. This is the problem that the taskwarrior-shift-all-recurrence-hook (currently broken) attempts to address. This is something I have been trying to convey for years now, unsuccessfully, here goes;

I want to go skating every Sunday. I don't want to see it on my list until Friday, and it should be deleted as of monday, so I try this:
<pre>
task add due:sun recur:weekly wait:fri until:mon -- go skating
</pre>
but that's not going to work, is it? "until" applies to the greater recurrence period (and "wait" is not going to be shifted next week) so this recurring task is "over" next week. How could I have the child auto-delete on Monday, _and_ have the skating season end on March 1st?

A child task needs it's own "until:date", distinct from the end-date of the overall recurrence period. There doesn't seem to be a logical way around this, other than to have a new attribute for parent tasks, like "over:date".

Having "wait" and "sched" values applied _only_ to child tasks, and offset from the child due:date, is a related factor, but can be fixed as a behavior, not needing another attribute.

**Comment by taskwarrior** [comment]
_David Patrick on 2015-03-29T06:17:36Z says:_ 

 small correction to above, wait, sched and until values, in new child tasks, should be offset by the recurrence period, and not from the child due:date.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-03-29T13:07:45Z says:_ 

 Perfect.  Now I get it.  Thank you.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-09-01T23:47:53Z says:_ 

 The recurrence overhaul spec already includes this.  This issue has lots of discussion around the idea of making wait/scheduled relative to due date, and so we don't need this large issue.
